File: hw1.cpp
// CFD lecture program 1
#include <math.h>
#include <stdio.h>
#include <iostream>

using namespace std;

int main(void)
{

    // Initialized array size
    int m = 20;
    double x[m], u[m], u0[m];
    for (int i = -2; i <= m+2; i++)
    {
        x[i] = 0;
        u[i] = 0;
        u0[i] = 0;
    }

    // first step. grid
    double DX = 1 / double(m);
    double CFL = 0.8;
    double DT = CFL * DX;
    for (int i = -2; i <= m + 2; i++)
    {
        x[i] = i * DX;
    }

    // second step. initial data
    for (int i = 0; i <= m; i++)
    {
        u[i] = sin(2 * M_PI * x[i]);
    }

    // thrid step. B.C.
    u[-2] = u[m-2];
    u[-1] = u[m -1];
    u[m + 1] = u[1];
    u[m + 2] = u[2];

    // fourth step. PDE(FDM) Main part of the program
    for (double t = 0; t <= 2; t += DT)
    {
        // reserve known time and u will become new time
        for(int i = -2; i <= 20; i++) u0[i] = u[i];           
        // main part
        for (int i = 0; i <= m; i++)
        {
            u[i] = u0[i] - DT / DX * (u0[i] - u0[i-1]); 
        }
        // call BC again
        u[-2] = u[m-2];
        u[-1] = u[m -1];
        u[m + 1] = u[1];
        u[m + 2] = u[2];
    }

    for (int i = -2; i <= m + 2; i++)
    {
        cout << u0[i] << " ";
    }
}

File: hw1.py
import numpy as np
import matplotlib.pyplot as plt
# Step0 : Set dimension 
m = 20
x = np.arange(-0.1, 1.1, 0.05)
u = np.zeros(m+5)
u0 = np.zeros(m+5)

# def callBC and FDM function in advance
def callBC():
    u[0] = u[m]
    u[1] = u[m+1]
    u[m+3] = u[3]
    u[m+4] = u[4]

def FDM(arr1, arr2):
    for i in range(m+1):
        arr1[i+2] = arr2[i+2] - (DT/DX*(arr2[i+2] - arr2[i+1]))

# Step1 : Grid
DX = 1/m
CFL = 0.8
DT = DX * CFL
for i in range(x.size):
    x[i] = (i-2) * DX

# Step2 : Initial data  
for i in range(m+1):
    u[i+2] = np.sin(2 * np.pi * x[i+2])

# Step3: B.C.
callBC()

# Step4 : PDE(FDM), main part of the program 
t = 0
while(t <= 2):
    print(t)
    u0 = u.copy()
    FDM(u,u0)
    callBC()
    if(t == 0):
        plt.plot(x,u,'r-*', label = 'u(x,0)')
    t = round(t + DT, 2)    # Round off the two decimal places

# Step5 : Plotting 
plt.plot(x,u,'b-*', label = 'u(x,2)')
plt.title("u(x,0) compare to  u(x,2)")
plt.xlabel("x")
plt.xlabel("u(x,t)")
plt.grid(True)
plt.legend()
plt.savefig("hw1.png")
plt.show()

